File: plugfn/core/src/core/workflow-engine.ts
import type {
  Workflow,
  WorkflowContext,
  WorkflowExecution,
  WorkflowStats,
  ListWorkflowsOptions,
  WorkflowStep,
} from '../types/workflow.js';
import { WorkflowStatus, WorkflowExecutionStatus } from '../types/workflow.js';
import type { Logger } from '../types/action.js';
import { WorkflowStorage } from '../storage/workflow-storage.js';
import { WebhookHandler } from '../webhooks/webhook-handler.js';
import type { TriggerHandler } from '../types/trigger.js';

interface PersistedWorkflowState {
  idempotencyKey?: string;
  completedStepIds: string[];
  retriable: boolean;
  resumed: boolean;
  failedStepId?: string;
  errorCode?: string;
}

interface WorkflowTriggerBinding {
  backend: 'webhook-handler';
  provider: string;
  event: string;
  handler: TriggerHandler;
}

export class WorkflowEngineError extends Error {
  readonly code:
    | 'WORKFLOW_TRIGGER_UNREGISTER_FAILED'
    | 'WORKFLOW_DURABILITY_UNSUPPORTED'
    | 'WORKFLOW_DEFINITION_INVALID';
  readonly status: number;
  readonly retryable = false;
  readonly details: Record<string, unknown>;

  constructor(
    code:
      | 'WORKFLOW_TRIGGER_UNREGISTER_FAILED'
      | 'WORKFLOW_DURABILITY_UNSUPPORTED'
      | 'WORKFLOW_DEFINITION_INVALID',
    message: string,
    details: Record<string, unknown> = {}
  ) {
    super(message);
    this.name = 'WorkflowEngineError';
    this.code = code;
    this.status = 409;
    this.details = details;
  }
}

export class WorkflowEngine {
  private readonly triggerBindings = new Map<string, WorkflowTriggerBinding>();
  private readonly executionLocks = new Map<string, Promise<void>>();

  constructor(
    private workflowStorage: WorkflowStorage,
    private webhookHandler: WebhookHandler,
    private logger: Logger
  ) {}

  async create(workflow: Omit<Workflow, 'id' | 'createdAt' | 'updatedAt'>): Promise<Workflow> {
    assertUniqueWorkflowStepIds(workflow.definition.steps);
    const created = await this.workflowStorage.create(workflow);

    if (workflow.status === WorkflowStatus.Enabled) {
      await this.registerTrigger(created);
    }

    this.logger.info(`Workflow created: ${created.name}`, { workflowId: created.id });
    return created;
  }

  async get(id: string): Promise<Workflow | null> {
    return this.workflowStorage.get(id);
  }

  async list(options?: ListWorkflowsOptions): Promise<Workflow[]> {
    return this.workflowStorage.list(options?.userId, options?.status);
  }

  async rehydrateEnabledTriggers(): Promise<{ registered: number; failed: number }> {
    const workflows = await this.workflowStorage.list(undefined, WorkflowStatus.Enabled);
    let registered = 0;
    let failed = 0;

    for (const workflow of workflows) {
      try {
        await this.registerTrigger(workflow);
        registered += 1;
      } catch (error) {
        failed += 1;
        this.logger.error('Workflow trigger rehydration failed', {
          workflowId: workflow.id,
          error,
        });
      }
    }

    return { registered, failed };
  }

  async enable(id: string): Promise<void> {
    const workflow = await this.workflowStorage.get(id);
    if (!workflow) {
      throw new Error(`Workflow ${id} not found`);
    }
    assertUniqueWorkflowStepIds(workflow.definition.steps);

    const updated = await this.workflowStorage.update(id, { status: WorkflowStatus.Enabled });
    await this.registerTrigger(updated);

    this.logger.info(`Workflow enabled: ${workflow.name}`, { workflowId: id });
  }

  async disable(id: string): Promise<void> {
    const workflow = await this.workflowStorage.get(id);
    if (!workflow) {
      throw new Error(`Workflow ${id} not found`);
    }

    const updated = await this.workflowStorage.update(id, { status: WorkflowStatus.Disabled });
    await this.unregisterTrigger(updated);

    this.logger.info(`Workflow disabled: ${workflow.name}`, { workflowId: id });
  }

  async delete(id: string): Promise<void> {
    const workflow = await this.workflowStorage.get(id);
    if (workflow && (workflow.status === WorkflowStatus.Enabled || this.triggerBindings.has(workflow.id))) {
      await this.unregisterTrigger(workflow);
    }

    await this.workflowStorage.delete(id);
    this.logger.info(`Workflow deleted`, { workflowId: id });
  }

  async execute(workflowId: string, triggerPayload: any): Promise<WorkflowExecution> {
    const workflow = await this.workflowStorage.get(workflowId);
    if (!workflow) {
      throw new Error(`Workflow ${workflowId} not found`);
    }

    if (workflow.status !== WorkflowStatus.Enabled) {
      throw new Error(`Workflow ${workflowId} is not enabled`);
    }
    assertUniqueWorkflowStepIds(workflow.definition.steps);

    const idempotencyKey = this.resolveIdempotencyKey(triggerPayload);
    const lockKey = idempotencyKey ? `${workflowId}:${idempotencyKey}` : undefined;
    const previousExecution = lockKey
      ? this.executionLocks.get(lockKey) ?? Promise.resolve()
      : Promise.resolve();
    let releaseExecutionLock: (() => void) | undefined;
    let queueTail: Promise<void> | undefined;
    if (lockKey) {
      const current = new Promise<void>((resolve) => {
        releaseExecutionLock = resolve;
      });
      queueTail = previousExecution.catch(() => {}).then(() => current);
      this.executionLocks.set(lockKey, queueTail);
    }

    await previousExecution.catch(() => {});

    try {
      const existingExecution = idempotencyKey
        ? await this.workflowStorage.findExecutionByIdempotencyKey(workflowId, idempotencyKey)
        : null;

      if (existingExecution && existingExecution.status === WorkflowExecutionStatus.Completed) {
        return existingExecution;
      }

      const resumed = Boolean(existingExecution);
      const execution =
        existingExecution ??
        (await this.workflowStorage.createExecution({
          workflowId,
          status: WorkflowExecutionStatus.Running,
          input: this.buildExecutionInput(triggerPayload, idempotencyKey),
        }));

      if (existingExecution) {
        await this.workflowStorage.updateExecution(existingExecution.id, {
          status: WorkflowExecutionStatus.Running,
          error: undefined,
        });
      }

      const executionState = this.readExecutionState(existingExecution ?? null);
      const completedStepIds = [...executionState.completedStepIds];
      const contextData = { ...executionState.data };

      const context: WorkflowContext = {
        workflowId,
        executionId: execution.id,
        userId: workflow.userId,
        trigger: {
          provider: workflow.definition.trigger.provider,
          event: workflow.definition.trigger.event,
          payload: triggerPayload,
        },
        data: contextData,
      };
      const checkpointStep = async (stepId: string): Promise<void> => {
        if (!completedStepIds.includes(stepId)) {
          completedStepIds.push(stepId);
        }
        await this.workflowStorage.updateExecution(execution.id, {
          status: WorkflowExecutionStatus.Running,
          output: this.buildExecutionOutput(context.data, {
            idempotencyKey,
            completedStepIds,
            retriable: false,
            resumed,
          }),
        });
      };

      const startedAtMs = execution.startedAt ? execution.startedAt.getTime() : Date.now();
      let activeStepId: string | undefined;

      try {
        if (workflow.definition.trigger.filter) {
          const shouldRun = workflow.definition.trigger.filter(context);
          if (!shouldRun) {
            await this.workflowStorage.updateExecution(execution.id, {
              status: WorkflowExecutionStatus.Completed,
              output: this.buildExecutionOutput(context.data, {
                idempotencyKey,
                completedStepIds,
                retriable: false,
                resumed,
              }),
              completedAt: new Date(),
              durationMs: Date.now() - startedAtMs,
            });

            const skippedExecution = await this.getExecutionById(workflowId, execution.id);
            return skippedExecution ?? execution;
          }
        }

        for (const step of workflow.definition.steps) {
          activeStepId = step.id;
          if (completedStepIds.includes(step.id)) {
            continue;
          }

          await this.executeStep(step, context, completedStepIds, checkpointStep);
          await checkpointStep(step.id);
        }

        await this.workflowStorage.updateExecution(execution.id, {
          status: WorkflowExecutionStatus.Completed,
          output: this.buildExecutionOutput(context.data, {
            idempotencyKey,
            completedStepIds,
            retriable: false,
            resumed,
          }),
          completedAt: new Date(),
          durationMs: Date.now() - startedAtMs,
        });

        this.logger.info(`Workflow executed successfully: ${workflow.name}`, {
          workflowId,
          executionId: execution.id,
          duration: Date.now() - startedAtMs,
          resumed,
        });

        const completedExecution = await this.getExecutionById(workflowId, execution.id);
        return completedExecution ?? execution;
      } catch (error) {
        const failure = normalizeWorkflowExecutionFailure(error);
        this.logger.error(`Workflow execution failed: ${workflow.name}`, {
          workflowId,
          executionId: execution.id,
          error: failure.error,
        });

        if (workflow.definition.errorHandlers) {
          for (const errorHandler of workflow.definition.errorHandlers) {
            try {
              await errorHandler.handler(failure.error, context);
            } catch (handlerError) {
              this.logger.error('Error handler failed', { handlerError });
            }
          }
        }

        await this.workflowStorage.updateExecution(execution.id, {
          status: WorkflowExecutionStatus.Failed,
          error: `${failure.code}:${failure.error.message}`,
          output: this.buildExecutionOutput(context.data, {
            idempotencyKey,
            completedStepIds,
            retriable: failure.retriable,
            resumed,
            failedStepId: activeStepId,
            errorCode: failure.code,
          }),
          completedAt: new Date(),
          durationMs: Date.now() - startedAtMs,
        });

        throw failure.error;
      }
    } finally {
      if (lockKey && releaseExecutionLock) {
        releaseExecutionLock();
        if (this.executionLocks.get(lockKey) === queueTail) {
          this.executionLocks.delete(lockKey);
        }
      }
    }
  }

  private async executeStep(
    step: WorkflowStep,
    context: WorkflowContext,
    completedStepIds: string[],
    checkpointStep: (stepId: string) => Promise<void>
  ): Promise<void> {
    switch (step.type) {
      case 'action': {
        const result = await step.action(context);
        Object.assign(context.data, result);
        break;
      }

      case 'filter': {
        const shouldContinue = step.condition(context);
        if (!shouldContinue) {
          throw new Error('Filter condition not met');
        }
        break;
      }

      case 'branch': {
        const condition = step.condition(context);
        const branch = condition ? step.then : step.else;
        if (branch) {
          for (const branchStep of branch) {
            if (completedStepIds.includes(branchStep.id)) {
              continue;
            }
            await this.executeStep(branchStep, context, completedStepIds, checkpointStep);
            await checkpointStep(branchStep.id);
          }
        }
        break;
      }

      case 'parallel': {
        const pendingActions = step.actions
          .map((action, index) => ({
            action,
            checkpointId: parallelActionCheckpointId(step.id, index),
          }))
          .filter(({ checkpointId }) => !completedStepIds.includes(checkpointId));
        const results = await Promise.allSettled(
          pendingActions.map(({ action }) => action(context))
        );

        for (const [index, result] of results.entries()) {
          if (result.status === 'fulfilled') {
            await checkpointStep(pendingActions[index].checkpointId);
          }
        }

        const failure = results.find(
          (result): result is PromiseRejectedResult => result.status === 'rejected'
        );
        if (failure) {
          throw failure.reason;
        }
        break;
      }

      case 'delay': {
        const ms = this.convertDelayToMs(step);
        throw new WorkflowEngineError(
          'WORKFLOW_DURABILITY_UNSUPPORTED',
          `delay step ${step.id} requires a durable scheduler`,
          {
            stepId: step.id,
            durationMs: ms,
          }
        );
      }

      default: {
        const unsupportedType = (step as { type?: unknown }).type;
        throw new Error(`Unknown workflow step type: ${String(unsupportedType)}`);
      }
    }
  }

  private convertDelayToMs(step: any): number {
    const { duration, unit = 'ms' } = step;

    switch (unit) {
      case 'ms':
        return duration;
      case 'seconds':
        return duration * 1000;
      case 'minutes':
        return duration * 60 * 1000;
      case 'hours':
        return duration * 60 * 60 * 1000;
      case 'days':
        return duration * 24 * 60 * 60 * 1000;
      default:
        return duration;
    }
  }

  async getStats(workflowId: string): Promise<WorkflowStats> {
    const executions = await this.workflowStorage.listExecutions(workflowId, 1000);

    const totalExecutions = executions.length;
    const successfulExecutions = executions.filter(
      (execution) => execution.status === WorkflowExecutionStatus.Completed
    ).length;
    const failedExecutions = executions.filter(
      (execution) => execution.status === WorkflowExecutionStatus.Failed
    ).length;

    const successRate = totalExecutions > 0 ? (successfulExecutions / totalExecutions) * 100 : 0;

    const durations = executions
      .filter((execution) => execution.durationMs)
      .map((execution) => execution.durationMs as number);
    const avgDuration = durations.length > 0
      ? durations.reduce((left, right) => left + right, 0) / durations.length
      : 0;

    const lastExecutedAt = executions.length > 0 ? executions[0].startedAt : undefined;

    return {
      totalExecutions,
      successfulExecutions,
      failedExecutions,
      successRate,
      avgDuration,
      lastExecutedAt,
    };
  }

  private async registerTrigger(workflow: Workflow): Promise<void> {
    assertUniqueWorkflowStepIds(workflow.definition.steps);
    if (this.triggerBindings.has(workflow.id)) {
      await this.unregisterTrigger(workflow, { allowMissing: true });
    }

    const handler: TriggerHandler = async (event: any) => {
      try {
        await this.execute(workflow.id, event);
      } catch (error) {
        this.logger.error(`Workflow trigger execution failed`, { workflow: workflow.id, error });
      }
    };

    this.webhookHandler.on(
      workflow.definition.trigger.provider,
      workflow.definition.trigger.event,
      handler
    );

    this.triggerBindings.set(workflow.id, {
      backend: 'webhook-handler',
      provider: workflow.definition.trigger.provider,
      event: workflow.definition.trigger.event,
      handler,
    });

    this.logger.info('Workflow trigger registered', {
      workflowId: workflow.id,
      backend: 'webhook-handler',
      provider: workflow.definition.trigger.provider,
      event: workflow.definition.trigger.event,
    });
  }

  private async unregisterTrigger(
    workflow: Workflow,
    options: { allowMissing?: boolean } = {}
  ): Promise<void> {
    const binding = this.triggerBindings.get(workflow.id);

    if (!binding) {
      this.logger.error('Workflow trigger unregister failed: missing active binding', {
        workflowId: workflow.id,
        provider: workflow.definition.trigger.provider,
        event: workflow.definition.trigger.event,
      });

      if (options.allowMissing) {
        return;
      }

      throw new WorkflowEngineError(
        'WORKFLOW_TRIGGER_UNREGISTER_FAILED',
        `active trigger binding not found for workflow ${workflow.id}`,
        {
          workflowId: workflow.id,
          provider: workflow.definition.trigger.provider,
          event: workflow.definition.trigger.event,
          backend: 'webhook-handler',
        }
      );
    }

    this.webhookHandler.off(binding.provider, binding.event, binding.handler);
    this.triggerBindings.delete(workflow.id);

    this.logger.info('Workflow trigger unregistered', {
      workflowId: workflow.id,
      backend: binding.backend,
      provider: binding.provider,
      event: binding.event,
    });
  }

  private buildExecutionInput(triggerPayload: any, idempotencyKey?: string): any {
    return {
      ...triggerPayload,
      __workflow: {
        idempotencyKey,
      },
    };
  }

  private buildExecutionOutput(data: Record<string, any>, state: PersistedWorkflowState): any {
    return {
      data,
      __workflow: {
        idempotencyKey: state.idempotencyKey,
        completedStepIds: [...state.completedStepIds],
        retriable: state.retriable,
        resumed: state.resumed,
        failedStepId: state.failedStepId,
        errorCode: state.errorCode,
      },
    };
  }

  private readExecutionState(execution: WorkflowExecution | null): {
    data: Record<string, any>;
    completedStepIds: string[];
  } {
    if (!execution || !execution.output || typeof execution.output !== 'object') {
      return {
        data: {},
        completedStepIds: [],
      };
    }

    const output = execution.output as Record<string, any>;
    const workflowState = output.__workflow as Record<string, any> | undefined;
    const completedStepIds = Array.isArray(workflowState?.completedStepIds)
      ? workflowState?.completedStepIds.filter((value: unknown): value is string => {
          return typeof value === 'string' && value.length > 0;
        })
      : [];

    const data = output.data && typeof output.data === 'object'
      ? (output.data as Record<string, any>)
      : {};

    return {
      data,
      completedStepIds,
    };
  }

  private resolveIdempotencyKey(triggerPayload: any): string | undefined {
    if (!triggerPayload || typeof triggerPayload !== 'object') {
      return undefined;
    }

    const explicit = triggerPayload.idempotencyKey;
    if (typeof explicit === 'string' && explicit.length > 0) {
      return explicit;
    }

    const payloadId = triggerPayload.id;
    if (typeof payloadId === 'string' && payloadId.length > 0) {
      return payloadId;
    }

    return undefined;
  }

  private async getExecutionById(
    workflowId: string,
    executionId: string
  ): Promise<WorkflowExecution | null> {
    const execution = await this.workflowStorage.getExecution(executionId);
    if (!execution || execution.workflowId !== workflowId) {
      return null;
    }

    return execution;
  }
}

function parallelActionCheckpointId(stepId: string, actionIndex: number): string {
  return `__parallel__:${stepId}:${actionIndex}`;
}

function assertUniqueWorkflowStepIds(steps: WorkflowStep[]): void {
  const seen = new Set<string>();
  const duplicates = new Set<string>();

  const visit = (step: WorkflowStep): void => {
    if (seen.has(step.id)) {
      duplicates.add(step.id);
    } else {
      seen.add(step.id);
    }

    if (step.type === 'branch') {
      for (const nested of [...step.then, ...(step.else ?? [])]) {
        visit(nested);
      }
    }
  };

  for (const step of steps) {
    visit(step);
  }

  if (duplicates.size > 0) {
    const duplicateStepIds = [...duplicates].sort((left, right) => left.localeCompare(right));
    throw new WorkflowEngineError(
      'WORKFLOW_DEFINITION_INVALID',
      `workflow step IDs must be unique: ${duplicateStepIds.join(', ')}`,
      { duplicateStepIds }
    );
  }
}

function normalizeWorkflowExecutionFailure(error: unknown): {
  code: string;
  retriable: boolean;
  error: Error;
} {
  if (error instanceof WorkflowEngineError) {
    return {
      code: error.code,
      retriable: false,
      error,
    };
  }

  if (error instanceof Error) {
    return {
      code: 'WORKFLOW_STEP_FAILED',
      retriable: true,
      error,
    };
  }

  return {
    code: 'WORKFLOW_STEP_FAILED',
    retriable: true,
    error: new Error('workflow execution failed'),
  };
}

File: plugfn/python/plugfn/core/workflow_engine.py
"""Workflow engine for managing and executing workflows."""

import asyncio
from datetime import datetime
from typing import Any, Awaitable, Callable, Dict, List, Optional, Tuple

from ..storage.workflow_storage import WorkflowStorage
from ..types import Workflow, WorkflowStatus


class WorkflowEngineError(RuntimeError):
    """Structured workflow lifecycle failure."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


class WorkflowEngine:
    """Engine for managing workflows."""

    def __init__(
        self,
        storage: WorkflowStorage,
        webhook_handler: Any,
        logger: Any,
        action_executor: Optional[Any] = None,
    ) -> None:
        """Initialize workflow engine.

        Args:
            storage: Workflow storage
            webhook_handler: Webhook handler for triggers
            logger: Logger instance
        """
        self.storage = storage
        self.webhook_handler = webhook_handler
        self.logger = logger
        self.action_executor = action_executor
        self._ready = False
        self._ready_lock = asyncio.Lock()
        self._trigger_bindings: Dict[
            str, Tuple[str, str, Callable[[Dict[str, Any]], Awaitable[Any]]]
        ] = {}

    async def list_workflows(
        self, user_id: Optional[str] = None, status: Optional[WorkflowStatus] = None
    ) -> List[Workflow]:
        """List workflows.

        Args:
            user_id: Optional user ID filter
            status: Optional status filter

        Returns:
            List of workflows
        """
        workflows = await self.storage.list_workflows(user_id=user_id)

        if status:
            workflows = [w for w in workflows if w.status == status]

        return workflows

    async def get_workflow(self, workflow_id: str) -> Optional[Workflow]:
        """Get a workflow by ID.

        Args:
            workflow_id: Workflow ID

        Returns:
            Workflow or None if not found
        """
        return await self.storage.get_workflow(workflow_id)

    async def create_workflow(
        self,
        user_id: str,
        name: str,
        definition: Dict[str, Any],
        description: Optional[str] = None,
    ) -> Workflow:
        """Create a new workflow.

        Args:
            user_id: User ID
            name: Workflow name
            definition: Workflow definition
            description: Optional description

        Returns:
            Created workflow
        """
        import secrets

        workflow_id = f"wf_{secrets.token_urlsafe(16)}"
        now = datetime.now()

        workflow = Workflow(
            id=workflow_id,
            user_id=user_id,
            name=name,
            description=description,
            definition=definition,
            status=WorkflowStatus.DRAFT,
            metadata={},
            created_at=now,
            updated_at=now,
        )

        await self.storage.create_workflow(workflow)

        self.logger.info(f"Created workflow: {workflow_id}", {"user_id": user_id})

        return workflow

    async def enable_workflow(self, workflow_id: str) -> None:
        """Enable a workflow.

        Args:
            workflow_id: Workflow ID

        Raises:
            ValueError: If workflow not found
        """
        workflow = await self.get_workflow(workflow_id)
        if not workflow:
            raise ValueError(f"Workflow {workflow_id} not found")

        self._trigger_coordinates(workflow)
        previous_binding = self._trigger_bindings.get(workflow.id)
        try:
            await self._register_trigger(workflow)
        except Exception:
            if previous_binding is not None:
                self._restore_trigger_binding(workflow.id, previous_binding)
            raise
        try:
            await self.storage.update_workflow(
                workflow_id,
                {"status": WorkflowStatus.ENABLED, "updated_at": datetime.now()},
            )
        except Exception:
            self._unregister_trigger(workflow, required=False)
            if previous_binding is not None:
                self._restore_trigger_binding(workflow.id, previous_binding)
            raise

        self.logger.info(f"Enabled workflow: {workflow_id}")

    async def disable_workflow(self, workflow_id: str) -> None:
        """Disable a workflow.

        Args:
            workflow_id: Workflow ID

        Raises:
            ValueError: If workflow not found
        """
        workflow = await self.get_workflow(workflow_id)
        if not workflow:
            raise ValueError(f"Workflow {workflow_id} not found")

        binding = self._unregister_trigger(
            workflow, required=workflow.status == WorkflowStatus.ENABLED
        )
        try:
            await self.storage.update_workflow(
                workflow_id,
                {"status": WorkflowStatus.DISABLED, "updated_at": datetime.now()},
            )
        except Exception:
            if binding is not None:
                self._restore_trigger_binding(workflow.id, binding)
            raise

        self.logger.info(f"Disabled workflow: {workflow_id}")

    async def delete_workflow(self, workflow_id: str) -> None:
        """Delete a workflow.

        Args:
            workflow_id: Workflow ID

        Raises:
            ValueError: If workflow not found
        """
        workflow = await self.get_workflow(workflow_id)
        if not workflow:
            raise ValueError(f"Workflow {workflow_id} not found")

        binding = self._unregister_trigger(
            workflow, required=workflow.status == WorkflowStatus.ENABLED
        )
        try:
            await self.storage.delete_workflow(workflow_id)
        except Exception:
            if binding is not None:
                self._restore_trigger_binding(workflow.id, binding)
            raise

        self.logger.info(f"Deleted workflow: {workflow_id}")

    async def _register_trigger(self, workflow: Workflow) -> None:
        provider, event = self._trigger_coordinates(workflow)

        self._unregister_trigger(workflow, required=False)

        async def handler(payload: Dict[str, Any]) -> Any:
            return await self.execute_workflow(workflow.id, payload)

        self.webhook_handler.register_handler(provider, event, handler)
        self._trigger_bindings[workflow.id] = (provider, event, handler)

    @staticmethod
    def _trigger_coordinates(workflow: Workflow) -> Tuple[str, str]:
        trigger = workflow.definition.get("trigger")
        if not isinstance(trigger, dict):
            raise ValueError(f"Workflow {workflow.id} trigger is invalid")

        provider = trigger.get("provider")
        event = trigger.get("event")
        if not isinstance(provider, str) or not provider:
            raise ValueError(f"Workflow {workflow.id} trigger provider is invalid")
        if not isinstance(event, str) or not event:
            raise ValueError(f"Workflow {workflow.id} trigger event is invalid")
        return provider, event

    def _restore_trigger_binding(
        self,
        workflow_id: str,
        binding: Tuple[str, str, Callable[[Dict[str, Any]], Awaitable[Any]]],
    ) -> None:
        provider, event, handler = binding
        self.webhook_handler.register_handler(provider, event, handler)
        self._trigger_bindings[workflow_id] = binding

    def _unregister_trigger(
        self, workflow: Workflow, *, required: bool
    ) -> Optional[Tuple[str, str, Callable[[Dict[str, Any]], Awaitable[Any]]]]:
        binding = self._trigger_bindings.pop(workflow.id, None)
        if binding is None:
            if required:
                raise WorkflowEngineError(
                    "WORKFLOW_TRIGGER_UNREGISTER_FAILED",
                    f"Workflow {workflow.id} has no registered trigger binding",
                )
            return None
        provider, event, handler = binding
        self.webhook_handler.unregister_handler(provider, event, handler)
        return binding

    async def rehydrate_enabled_triggers(self) -> Dict[str, int]:
        """Restore persisted enabled trigger bindings after process startup."""
        workflows = await self.storage.list_workflows(status=WorkflowStatus.ENABLED)
        registered = 0
        failed = 0
        for workflow in workflows:
            try:
                await self._register_trigger(workflow)
                registered += 1
            except Exception as error:
                failed += 1
                self.logger.error(
                    f"Failed to rehydrate workflow trigger: {workflow.id}",
                    {"error": str(error)},
                )
        return {"registered": registered, "failed": failed}

    async def ready(self) -> None:
        """Ensure enabled triggers are restored; failed attempts remain retryable."""
        if self._ready:
            return
        async with self._ready_lock:
            if self._ready:
                return
            result = await self.rehydrate_enabled_triggers()
            if result["failed"]:
                raise WorkflowEngineError(
                    "WORKFLOW_TRIGGER_REHYDRATION_FAILED",
                    f"Failed to rehydrate {result['failed']} workflow trigger(s)",
                )
            self._ready = True

    async def get_workflow_stats(self, workflow_id: str) -> Dict[str, Any]:
        """Get workflow execution statistics.

        Args:
            workflow_id: Workflow ID

        Returns:
            Workflow statistics

        Raises:
            ValueError: If workflow not found
        """
        workflow = await self.get_workflow(workflow_id)
        if not workflow:
            raise ValueError(f"Workflow {workflow_id} not found")

        # Get executions
        executions = await self.storage.list_workflow_executions(workflow_id)

        total = len(executions)
        completed = len([e for e in executions if e.get("status") == "completed"])
        failed = len([e for e in executions if e.get("status") == "failed"])

        avg_duration = 0
        if executions:
            durations = [e.get("duration", 0) for e in executions if e.get("duration")]
            if durations:
                avg_duration = sum(durations) / len(durations)

        return {
            "workflow_id": workflow_id,
            "total_executions": total,
            "completed": completed,
            "failed": failed,
            "success_rate": completed / total if total > 0 else 0,
            "avg_duration": avg_duration,
        }

    async def execute_workflow(
        self, workflow_id: str, trigger_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Execute a workflow.

        Args:
            workflow_id: Workflow ID
            trigger_data: Trigger event data

        Returns:
            Execution result

        Raises:
            ValueError: If workflow not found or disabled
        """
        workflow = await self.get_workflow(workflow_id)
        if not workflow:
            raise ValueError(f"Workflow {workflow_id} not found")

        if workflow.status != WorkflowStatus.ENABLED:
            raise ValueError(f"Workflow {workflow_id} is not enabled")

        # Create execution record
        import secrets

        execution_id = f"exec_{secrets.token_urlsafe(16)}"
        start_time = datetime.now()

        execution = {
            "id": execution_id,
            "workflow_id": workflow_id,
            "status": "running",
            "trigger_data": trigger_data,
            "started_at": start_time,
            "steps": [],
        }

        await self.storage.create_workflow_execution(execution)

        # Execute workflow steps (simplified)
        # In a full implementation, this would process the workflow definition
        try:
            # Process workflow definition
            steps = workflow.definition.get("steps", [])
            if not isinstance(steps, list):
                raise ValueError(f"Workflow {workflow_id} steps are invalid")
            results = []

            for step in steps:
                if not isinstance(step, dict) or step.get("type") != "action":
                    raise ValueError("Workflow step type must be 'action'")
                if self.action_executor is None:
                    raise RuntimeError("Workflow action executor is not configured")
                provider = step.get("provider")
                action = step.get("action")
                params = step.get("params", {})
                if not isinstance(provider, str) or not provider:
                    raise ValueError("Workflow action step provider is invalid")
                if not isinstance(action, str) or not action:
                    raise ValueError("Workflow action step action is invalid")
                if not isinstance(params, dict):
                    raise ValueError("Workflow action step params are invalid")

                action_result = await self.action_executor.execute(
                    provider=provider,
                    action=action,
                    user_id=workflow.user_id,
                    params=params,
                    connection_id=step.get("connection_id"),
                )
                if action_result.get("success") is False:
                    raise RuntimeError(str(action_result.get("error") or "Workflow action failed"))
                step_result = {
                    "step": step,
                    "status": "completed",
                    "result": action_result,
                }
                results.append(step_result)

            # Update execution
            end_time = datetime.now()
            duration = int((end_time - start_time).total_seconds() * 1000)

            await self.storage.update_workflow_execution(
                execution_id,
                {
                    "status": "completed",
                    "steps": results,
                    "completed_at": end_time,
                    "duration": duration,
                },
            )

            return {
                "execution_id": execution_id,
                "status": "completed",
                "duration": duration,
                "results": results,
            }

        except Exception as e:
            # Update execution with error
            end_time = datetime.now()
            duration = int((end_time - start_time).total_seconds() * 1000)

            await self.storage.update_workflow_execution(
                execution_id,
                {
                    "status": "failed",
                    "error": str(e),
                    "completed_at": end_time,
                    "duration": duration,
                },
            )

            raise
